clippy_lints/src/lifetimes.rs

```diff
@@ -301,47 +318,62 @@ fn could_use_elision<'tcx>(
         for lt in input_visitor.nested_elision_site_lts {
             if let RefLt::Named(def_id) = lt {
                 if allowed_lts.contains(&cx.tcx.item_name(def_id.to_def_id())) {
-                    return false;
+                    return None;
                 }
             }
         }
         for lt in output_visitor.nested_elision_site_lts {
             if let RefLt::Named(def_id) = lt {
                 if allowed_lts.contains(&cx.tcx.item_name(def_id.to_def_id())) {
-                    return false;
+                    return None;
                 }
             }
         }
     }
 
     // no input lifetimes? easy case!
     if input_lts.is_empty() {
-        false
+        None
     } else if output_lts.is_empty() {
         // no output lifetimes, check distinctness of input lifetimes
 
         // only unnamed and static, ok
         let unnamed_and_static = input_lts.iter().all(|lt| *lt == RefLt::Unnamed || *lt == RefLt::Static);
         if unnamed_and_static {
-            return false;
+            return None;
+        }
+        // we have no output reference, so we can elide explicit lifetimes that occur at most once
+        let elidable_lts = named_lifetime_occurrences(&input_lts)
+            .into_iter()
+            .filter_map(|(def_id, occurrences)| {
+                if occurrences <= 1 {
+                    Some((def_id, input_visitor.sample_generic_arg_span.get(&def_id).copied()))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        if elidable_lts.is_empty() {
+            None
+        } else {
+            Some(elidable_lts)
         }
-        // we have no output reference, so we only need all distinct lifetimes
-        input_lts.len() == unique_lifetimes(&input_lts)
     } else {
         // we have output references, so we need one input reference,
         // and all output lifetimes must be the same
-        if unique_lifetimes(&output_lts) > 1 {
-            return false;
-        }
```

Comment:
This was a mistake.

I have significantly reworked this code.

If I am thinking about this right:
```
    // A lifetime can be newly elided if:
    // - It occurs only once among the inputs.
    // - If there are multiple input lifetimes, then the newly elided lifetime does not occur among the
    //   outputs (because eliding such an lifetime would create an ambiguity).
```
The changes produce several new warnings.

I apologize for the mistake and for the significance of the changes.

Reply:
That sounds good to me

I believe `&'a self`/`&'a mut self` would be an exception to the second point, but that can be a future addition

Also a heads up if you're testing locally there's [an elision bug](https://github.com/rust-lang/rust/issues/103330) that was fixed in the latest nightly (not yet in clippy)

No worries about a mistake, I wasn't sure if it was the comment or the code that could do with a change

Reply:
> I believe `&'a self`/`&'a mut self` would be an exception to the second point, but that can be a future addition

To be sure I understand, you're saying that additional elisions would be possible, not that this PR would introduce false positives, correct?

> Also a heads up if you're testing locally there's [an elision bug](https://github.com/rust-lang/rust/issues/103330) that was fixed in the latest nightly (not yet in clippy)

:+1:

Reply:
Yeah, e.g. `fn f<'a, 'b>(&'a self, _: &'b str) -> &'a str` will currently tell you `'b` can be elided, but it could also suggest eliding `'a`

Reply:
I would be happy to look into that for a subsequent PR.